File: uber/src/main/java/com/minageorge/uber/UberApplication.java
package com.minageorge.uber;

import android.app.Application;
import android.content.Context;

import com.minageorge.uber.di.application.AppComponent;
import com.minageorge.uber.di.application.AppModule;
import com.minageorge.uber.di.application.DaggerAppComponent;

public class UberApplication extends Application {

    private static Context context;
    private final AppComponent appComponent = createAppComponent();

    public static AppComponent getComponent(Context context) {
        return getApp(context).appComponent;
    }

    //This is a hack to get a non-static field from a static method (i.e. appComponent)
    private static UberApplication getApp(Context context) {
        return (UberApplication) context.getApplicationContext();
    }

    public static Context getContext() {
        return context;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        context = this;
    }

    private AppComponent createAppComponent() {
        return DaggerAppComponent.builder()
                .appModule(new AppModule(this))
                .build();
    }

}

File: uber/src/main/java/com/minageorge/uber/UberApplication.kt
package com.minageorge.uber

import android.app.Application
import com.minageorge.uber.store.room.UberDataBase
import com.minageorge.uber.ui.home.HomeViewModelFactory
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.android.x.androidXModule
import org.kodein.di.generic.bind
import org.kodein.di.generic.instance
import org.kodein.di.generic.provider
import org.kodein.di.generic.singleton


class UberApplication : Application(), KodeinAware {
    override val kodein = Kodein.lazy {
        import(androidXModule(this@UberApplication))
        bind() from singleton { UberDataBase(instance()) }
        bind() from provider { HomeViewModelFactory(instance(),instance()) }
    }

}